src/AspNet.Security.OAuth.VisualStudio/VisualStudioAuthenticationHandler.cs

```diff
@@ -66,7 +63,16 @@ protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OA
             ["client_assertion_type"] = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer"
         };
 
-        request.Content = new FormUrlEncodedContent(parameters!);
+        // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+        if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+        {
+            tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+            context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        }
+
+        using var request = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/x-www-form-urlencoded"));
```

Comment:
I see it's not a change introduced in this PR but are we sure we use the right value here? The OAuth 2.0 specification requires returning JSON responses and we use `JsonDocument.Parse()` later in the method, so it's not clear to me why formurl encoding is used here. Just a mistake?

Reply:
Makes sense 👍🏻